pyjobshop/solvers/ortools/Constraints.py

```diff
@@ -128,14 +121,58 @@ def _resource_breaks_constraints(self):
         """
         model, data, variables = self._model, self._data, self._variables
 
-        for idx, resource in enumerate(data.resources):
-            break_intervals = [
-                model.new_fixed_size_interval_var(start, end - start, "")
-                for start, end in resource.breaks
-            ]
+        for task_idx, task in enumerate(data.tasks):
+            if task.allow_breaks:
+                continue
+
+            for res_idx in data.task2resources(task_idx):
+                # Assignment variables should not overlap with any break. Using
+                # the `no_overlap` also handles the case when the assignment
+                # variable is absent.
+                assign_var = variables.assign_vars[task_idx, res_idx]
+                break_intervals = [
+                    model.new_fixed_size_interval_var(start, end - start, "")
+                    for start, end in data.resources[res_idx].breaks
+                ]
+                model.add_no_overlap([assign_var.interval, *break_intervals])
+
+        for task_idx, task in enumerate(data.tasks):
+            if not task.allow_breaks:
+                continue
+
+            task_var = variables.task_vars[task_idx]
+
+            for mode_idx in data.task2modes(task_idx):
+                mode_var = variables.mode_vars[mode_idx]
+                overlap_vars = variables.overlap_vars[mode_idx]
+
+                # Set the task break duration equal to the total overlap of
+                # the selected mode variable.
+                total_overlap = sum(var.duration for var in overlap_vars)
+                expr = task_var.breaks == total_overlap
+                model.add(expr).only_enforce_if(mode_var)
 
-            for var in variables.res2assign(idx):
-                model.add_no_overlap([var.interval, *break_intervals])
+                # Cannot start or end during a break. The domains below capture
+                # the invalid start/end times, and the complement ensures that
+                # these values are excluded.
+                breaks = [var.break_time for var in overlap_vars]
+                starts = Domain.from_intervals([[s, e - 1] for s, e in breaks])
```

Comment:
This affects performance significantly. May need to remove this and post-process.